### PR TITLE
Animal User Story

### PR DESCRIPTION
**Pull request by Mononofuu**
- Animal class, Dto, table  reworked
- Species class, Dto, table  reworked
- JdbcTemplate persistence provider  added
- Animal, Species  and House specifications added
- House capacity map initialization  added to House Repository
- House and Animal business logic in services
- House exception
- Animal exception

### CODE REVIEW

File: persistence/src/main/java/edu/softserve/zoo/persistence/provider/JdbcTemplatePersistenceProvider.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+    /**
+     * Get parametrized map from the persistent storage according to given Specification.
+     * In most keys key is an Id and value is a result of an aggregation function
+     *
+     * @param specification defines restrictions for performed search.
+     * @return map of Key-Values pairs
+     * @throws PersistenceProviderException if the received specification is null
+     */
+    <K extends Number, V extends Number> Map<K, V> getMap(SqlGetMapSpecification<K, V> specification);
```
Comment (ghost) [inline persistence/src/main/java/edu/softserve/zoo/persistence/provider/JdbcTemplatePersistenceProvider.java:23]: why do you narrow generics to Numbers? what's for?

File: persistence/src/main/java/edu/softserve/zoo/persistence/repository/impl/HouseRepositoryImpl.java [hunk 2]
```diff
@@ -1,16 +1,39 @@
...
 public class HouseRepositoryImpl extends AbstractRepository<House> implements HouseRepository {
+
+    @Autowired
+    private JdbcTemplatePersistenceProvider persistenceProvider;
+
+    private Map<Long, Long> houseCapacityMap;
+
+    @PostConstruct
+    public void initHouseCapacityMap() {
+        houseCapacityMap = Collections.synchronizedMap(persistenceProvider.getMap(new HouseGetCapacityMapSpecification()));
```
Comment (ghost) [inline persistence/src/main/java/edu/softserve/zoo/persistence/repository/impl/HouseRepositoryImpl.java]: bad practice to use collections.syncMap. look at thread safe variants

Comment (ghost) [inline persistence/src/main/java/edu/softserve/zoo/persistence/repository/impl/HouseRepositoryImpl.java]: why do you need sync map?

File: persistence/src/main/java/edu/softserve/zoo/persistence/specification/hibernate/impl/house/HouseGetAllBySpeciesIdSpecification.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * @author Serhii Alekseichenko
+ */
+public class HouseGetAllBySpeciesIdSpecification implements HQLSpecification<House> {
+    private final Long speciesId;
+
+    /**
+     * @param speciesId {@link Species} identifier to filter houses
+     */
+    public HouseGetAllBySpeciesIdSpecification(Long speciesId) {
+        this.speciesId = speciesId;
```
Comment (ghost) [inline persistence/src/main/java/edu/softserve/zoo/persistence/specification/hibernate/impl/house/HouseGetAllBySpeciesIdSpecification.java:26]: ok. assume this id is null. what's next?

File: persistence/src/main/java/edu/softserve/zoo/persistence/provider/JdbcTemplatePersistenceProvider.java (hunk 1 above)
Comment (Mononofuu) [inline persistence/src/main/java/edu/softserve/zoo/persistence/provider/JdbcTemplatePersistenceProvider.java:23]: Maybe the main reason - is that I consider this functional as a tool to return the ratio of the result of the aggregate function to id. Can you give me some advice on this?

File: persistence/src/main/java/edu/softserve/zoo/persistence/specification/hibernate/impl/house/HouseGetAllBySpeciesIdSpecification.java (hunk 3 above)
Comment (Mononofuu) [inline persistence/src/main/java/edu/softserve/zoo/persistence/specification/hibernate/impl/house/HouseGetAllBySpeciesIdSpecification.java:26]: validation was moved to the constructor, so exception will be thrown at this stage

**Pull request merged**